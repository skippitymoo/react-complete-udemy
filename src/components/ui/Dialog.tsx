import React, { JSX, forwardRef, useImperativeHandle, useRef } from 'react';
import { createPortal } from 'react-dom';

export type DialogProps = React.ComponentPropsWithRef<'dialog'>;

export type DialogRef = {
  /**
   * Close the dialog
   */
  close: () => void;
  /**
   * Open the dialog
   */
  open: () => void;
};

const Dialog = forwardRef<DialogRef, DialogProps>(function CustomDialog(
  { children },
  ref,
): JSX.Element {
  const dialog = useRef<HTMLDialogElement>(null);

  useImperativeHandle(ref, () => {
    return {
      open() {
        dialog?.current?.showModal();
      },
      close() {
        dialog?.current?.close();
      },
    };
  });

  return createPortal(
    <dialog ref={dialog} className='dialog'>
      {children}
    </dialog>,
    document?.getElementById('modal')!,
  );
});

export { Dialog };
