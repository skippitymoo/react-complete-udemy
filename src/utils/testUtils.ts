/**
 * React Testing Library uses jsdom, which currently does not support HTMLDialogElement
 * https://github.com/testing-library/react-testing-library/issues/1106#issuecomment-1209419325
 * https://github.com/jsdom/jsdom/issues/3294#issuecomment-1268330372
 */
export const setUpModalContainer = (): HTMLDivElement => {
  document.body.replaceChildren();
  const modalDiv = document.createElement('div');
  modalDiv.setAttribute('id', 'modal');
  const modalContainer = document.body.appendChild(modalDiv);

  const showMock = jest.fn().mockImplementation(function mock(this: HTMLDialogElement) {
    this.setAttribute('open', '');
    this.open = true;
  });

  HTMLDialogElement.prototype.show = showMock;

  HTMLDialogElement.prototype.showModal = showMock;

  HTMLDialogElement.prototype.close = jest.fn().mockImplementation(function mock(
    this: HTMLDialogElement,
  ) {
    this.open = false;
  });

  return modalContainer;
};
