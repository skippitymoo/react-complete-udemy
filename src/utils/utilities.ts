export const formatCurrency = (val: number): string => {
  return new Intl.NumberFormat('en-GB', {
    style: 'currency',
    currency: 'GBP',
  }).format(val);
};
