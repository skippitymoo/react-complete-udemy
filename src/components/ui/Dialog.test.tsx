import { act, render, screen, waitFor } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import React from 'react';
import { Dialog, DialogRef } from './Dialog';
import { setUpModalContainer } from '../../utils/testUtils';

expect.extend(toHaveNoViolations);

describe('Dialog', () => {
  let modalContainer: HTMLDivElement;

  beforeEach(() => {
    modalContainer = setUpModalContainer();
  });

  it('has no accessibility violations', async () => {
    const { container } = render(<Dialog>dummy</Dialog>, { container: modalContainer });
    expect(await axe(container)).toHaveNoViolations();
  });

  it('should render', async () => {
    const dialogRef = React.createRef<DialogRef>();

    const { container } = render(
      <Dialog ref={dialogRef} className='random-class'>
        <div data-testid='my-text'>yeooo !</div>
      </Dialog>,
      { container: modalContainer },
    );

    act(() => {
      dialogRef.current?.open();
    });

    await waitFor(() => {
      expect(screen.getByTestId('my-text')).toHaveTextContent('yeooo !');
    });

    expect(container).toMatchSnapshot();
  });

  it('should be visible when dialog open', async () => {
    const dialogRef = React.createRef<DialogRef>();

    const { container } = render(
      <Dialog ref={dialogRef}>
        <div>yeooo !</div>
      </Dialog>,
      { container: modalContainer },
    );

    act(() => {
      dialogRef.current?.open();
    });

    await waitFor(() => {
      expect(screen.getByText(/yeooo !/i)).toBeVisible();
    });

    expect(container).toMatchSnapshot();
  });

  it('should not be visible when dialog closed', async () => {
    const dialogRef = React.createRef<DialogRef>();

    const { container } = render(
      <Dialog ref={dialogRef}>
        <div>yeooo !</div>
      </Dialog>,
      { container: modalContainer },
    );

    act(() => {
      dialogRef.current?.open();
    });

    await waitFor(() => {
      expect(screen.getByText(/yeooo !/i)).toBeVisible();
    });

    act(() => {
      dialogRef.current?.close();
    });

    await waitFor(() => {
      expect(screen.getByText(/yeooo !/i)).not.toBeVisible();
    });

    expect(container).toMatchSnapshot();
  });

  it('should initially open dialog', async () => {
    render(
      <Dialog initShowModal={true}>
        <div>yeooo !</div>
      </Dialog>,
      { container: modalContainer },
    );

    await waitFor(() => {
      expect(screen.getByText(/yeooo !/i)).toBeVisible();
    });
  });
});
