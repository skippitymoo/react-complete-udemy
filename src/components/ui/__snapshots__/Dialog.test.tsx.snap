// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should be visible when dialog open 1`] = `
<div
  id="modal"
>
  <dialog
    class="dialog"
    open=""
  >
    <div>
      yeooo !
    </div>
  </dialog>
</div>
`;

exports[`Dialog should not be visible when dialog closed 1`] = `
<div
  id="modal"
>
  <dialog
    class="dialog"
  >
    <div>
      yeooo !
    </div>
  </dialog>
</div>
`;

exports[`Dialog should render 1`] = `
<div
  id="modal"
>
  <dialog
    class="dialog random-class"
    open=""
  >
    <div
      data-testid="my-text"
    >
      yeooo !
    </div>
  </dialog>
</div>
`;
